tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,638 @@
+    const StorageT quantization_min, const StorageT quantization_max,
+    const absl::Span<const StorageT> input_zero_points,
+    const absl::Span<const ExpressedT> input_scales, const Strides& strides,
+    const StorageT* input_data, ExpressedT* input_dequantized_data,
+    const size_t depth, size_t quantization_index) {
+  const DimensionSize dim = shape.Dim(depth);
+  if (depth + 1 >= shape.Rank()) {
+    for (DimensionSize i = 0; i < dim; ++i) {
+      if (depth == quantization_dimension) {
+        quantization_index = i;
+      }
+      *input_dequantized_data =
+          Dequantize(*input_data, input_zero_points[quantization_index],
+                     input_scales[quantization_index]);
+      input_data += strides[depth];
+      input_dequantized_data += strides[depth];
+    }
+  } else {
+    for (DimensionSize i = 0; i < dim; ++i) {
+      if (depth == quantization_dimension) {
+        quantization_index = i;
+      }
+      DequantizeOpQuantizePerAxisImpl(
+          shape, quantization_dimension, quantization_min, quantization_max,
+          input_zero_points, input_scales, strides, input_data,
+          input_dequantized_data, depth + 1, quantization_index);
+      input_data += strides[depth];
+      input_dequantized_data += strides[depth];
+    }
+  }
+}
+
+template <typename StorageT, typename ExpressedT>
+void QuantizeOpQuantizePerAxisImpl(
+    const Shape& shape, const Axis quantization_dimension,
+    const StorageT quantization_min, const StorageT quantization_max,
+    const absl::Span<const StorageT> input_zero_points,
+    const absl::Span<const ExpressedT> input_scales, const Strides& strides,
+    StorageT* input_data, ExpressedT* input_dequantized_data,
+    const size_t depth, size_t quantization_index) {
+  const DimensionSize dim = shape.Dim(depth);
+  if (depth + 1 >= shape.Rank()) {
+    for (DimensionSize i = 0; i < dim; ++i) {
+      if (depth == quantization_dimension) {
+        quantization_index = i;
+      }
+      *input_data = Quantize<StorageT, ExpressedT>(
+          *input_dequantized_data, input_zero_points[quantization_index],
+          static_cast<ExpressedT>(1 / input_scales[quantization_index]),
+          quantization_min, quantization_max);
+      input_data += strides[depth];
+      input_dequantized_data += strides[depth];
+    }
+  } else {
+    for (DimensionSize i = 0; i < dim; ++i) {
+      if (depth == quantization_dimension) {
+        quantization_index = i;
+      }
+      QuantizeOpQuantizePerAxisImpl(
+          shape, quantization_dimension, quantization_min, quantization_max,
+          input_zero_points, input_scales, strides, input_data,
+          input_dequantized_data, depth + 1, quantization_index);
+      input_data += strides[depth];
+      input_dequantized_data += strides[depth];
+    }
+  }
+}
+
+template <DataType storage_type, DataType expressed_type>
+void DequantizeOpQuantizePerAxis(DotGeneralOp& op, const Tensor& lhs,
```

Comment:
Similar comment here about dequantizing the whole tensors. When quantized tensors are large, this becomes quite memory expensive.

Reply:
Now this calculations are done in the prepare step and not in the eval.